vsintegration/tests/FSharp.Editor.Tests/Refactors/RemoveExplicitReturnType.fs

```diff
@@ -0,0 +1,98 @@
+﻿module FSharp.Editor.Tests.Refactors.RemoveExplicitReturnType
+
+open Microsoft.VisualStudio.FSharp.Editor
+open Xunit
+open NUnit.Framework
+open FSharp.Editor.Tests.Refactors.RefactorTestFramework
+open Microsoft.CodeAnalysis.Text
+open Xunit
+open System.Runtime.InteropServices
+
+[<Theory>]
+[<InlineData(":int")>]
+[<InlineData(" :int")>]
+[<InlineData(" : int")>]
+[<InlineData(" :    int")>]
+let ``Refactor removes explicit return type`` (toRemove: string) =
+    task {
+        let symbolName = "sum"
+
+        let code =
+            $"""
+            let sum a b {toRemove}= a + b
+            """
+
+        use context = TestContext.CreateWithCode code
+        let spanStart = code.IndexOf symbolName
+
+        let! newDoc = tryRefactor code spanStart context (new RemoveExplicitReturnType())
+
+        do! AssertHasNoExplicitReturnType symbolName newDoc context.CT
+    }
+
+[<Fact>]
+let ``Refactor should not be suggested if theres nothing to remove`` () =
+    task {
+
+        let symbolName = "sum"
+
+        let code =
+            $"""
+            let sum a b = a + b
+            """
+
+        use context = TestContext.CreateWithCode code
+
+        let spanStart = code.IndexOf symbolName
+
+        let! actions = tryGetRefactoringActions code spanStart context (new RemoveExplicitReturnType())
+
+        do Assert.Empty(actions)
+    }
+
+[<Fact>]
+let ``Refactor should not be suggested if it changes the return type`` () =
+    task {
+        let symbolName = "sum"
+
+        let code =
```

Comment:
@psfinaki A bit of a roadblock here.  also see RemoveExplicitReturnType.fs Line 72. I dont see a way of checking, wether I can safely remove the return type other than completly copying the solution, remove the explicit return type and check if its still the same type. That seems a bit much?

Reply:
Mhmm let me play around with it - but this approach can probably go as the first take here, we are creating ad-hoc solutions and projects everywhere in the tests anyway.

As in, let's focus on how assertions should like for now and then we can improve the internal mechanics.

Reply:
A few more notes on testing:
1. It would be most expressive to have the tests sync, feel free to return `task.Result` from the refactor framework
2. Again about expressiveness, we tend to follow the pattern of having "expected code after applying the action" versus "actual code after applying the action". 

So it would be:
- initial code:
```fsharp
type B = { X:int }
type A = { X:int }
  
let f (i: int) : A = { X = i }
let sum a b = a + b
```
- desired code after the change:
```fsharp
type B = { X:int }
type A = { X:int }

let f (i: int) = { X = i }
let sum a b = a + b
```
- trigger the refactoring, get the actual changed document
- compare with expectations

Reply:
... and if you manage to intercept the changed code from the code refactoring to the test framework, you won't have to create extra projects or solutions. This approach works for code fixes as of now.

Reply:
@psfinaki Sorry, I may have not been clear enough. I mean in the refactor code. to know whether or not I can even offer the refactoring, I´d have to temporarily apply the refactoring and see if the type changed. Unless I´m missing something.

Otherwise, no problem, I make the test sync and compare actual documents. I may have overthunk a bit on that part XD

Reply:
Oh I see now, thanks for the clarification. Okay yeah that might be a bit heavy. Which is not the end of the world - we can always have this as an optional functionality. 

But just to clarify, do you think there are cases when we can be sure that the refactoring is going to work without "pre-applying" these changes? So that we can align with other refactorings we have already?

Reply:
> comparing the test doesn't actually evaluate whether the return type of "f" changes to B

Can you clarify on this as well? For testing, it is enough to know what changes the corresponding refactoring is going to apply and then apply them to the source text within the test framework itself - and then to compare to the expected result.

Reply:
@psfinaki Its about your (correct) observation that sometimes removing the explicit return type could corrupt the code : https://github.com/dotnet/fsharp/issues/15562#issuecomment-1752782017 . In that case, the refactoring should ideally not change the correctness. I keep researching , maybe there is still a way to determine other possible types for the symbol or something.

On the other side, just comparing the expected text output doesnt give me confidence that I dont accidently change the type of a symbol with the refactoring. But I can keep the test to that specific case. 

Considering all of that, maybe its time to split the Add and Remove Explicit Return type into seperate Issues. For Add, the text output is enough to test it and my logic is enough to correctly suggest the refactoring to the user.

Reply:
@psfinaki I saved the RemovExplicitReturnType stuff on a seperate branch and removed it.

Reply:
Alright, yeah let's go with the adding the explicit type for now and then add removing type as a followup - looks like it will be an interesting twist on this one :)